File: backend/schemas/diff.py
from datetime import datetime

from pydantic import BaseModel


class DiffRequest(BaseModel):
    baseline_id: int
    target_id: int
    diff_type: str = "full"   # full | findings | permissions | libraries


class FindingSnap(BaseModel):
    category: str
    severity: str
    title: str
    description: str
    file_path: str | None = None
    rule_id: str | None = None


class DiffOut(BaseModel):
    id: int
    created_at: datetime
    baseline_id: int | None
    target_id: int | None
    diff_type: str
    added_findings: list[FindingSnap]
    removed_findings: list[FindingSnap]
    added_permissions: list[str]
    removed_permissions: list[str]
    severity_delta: dict[str, int]   # e.g. {"critical": 2, "high": -1}
    summary: str | None

    model_config = {"from_attributes": True}


class DiffSummary(BaseModel):
    id: int
    created_at: datetime
    baseline_id: int | None
    target_id: int | None
    diff_type: str
    summary: str | None

    model_config = {"from_attributes": True}

File: frontend/src/api/diff.ts
import { api } from './client'

export interface FindingSnap {
  category: string
  severity: string
  title: string
  description: string
  file_path: string | null
  rule_id: string | null
}

export interface DiffOut {
  id: number
  created_at: string
  baseline_id: number | null
  target_id: number | null
  diff_type: string
  added_findings: FindingSnap[]
  removed_findings: FindingSnap[]
  added_permissions: string[]
  removed_permissions: string[]
  severity_delta: Record<string, number>
  summary: string | null
}

export interface DiffSummary {
  id: number
  created_at: string
  baseline_id: number | null
  target_id: number | null
  diff_type: string
  summary: string | null
}

export const diffApi = {
  create: (baseline_id: number, target_id: number, diff_type = 'full') =>
    api.post<DiffOut>('/diff', { baseline_id, target_id, diff_type }).then((r) => r.data),

  list: () => api.get<DiffSummary[]>('/diff').then((r) => r.data),

  get: (id: number) => api.get<DiffOut>(`/diff/${id}`).then((r) => r.data),

  delete: (id: number) => api.delete(`/diff/${id}`),
}
